CRM/Contribute/Tokens.php

```diff
@@ -96,35 +115,122 @@ public function alterActionScheduleQuery(\Civi\ActionSchedule\Event\MailingQuery
       $e->query->select("e." . $fields[$orig]['name'] . " AS contrib_{$alias}");
     }
   }
 
   /**
    * @inheritDoc
    */
-  public function evaluateToken(\Civi\Token\TokenRow $row, $entity, $field, $prefetch = NULL) {
-    $actionSearchResult = $row->context['actionSearchResult'];
-    $fieldValue = $actionSearchResult->{"contrib_$field"} ?? NULL;
+  public function prefetch(TokenValueEvent $e) {
+    // Find all the entity IDs
+    $entityIds
+      = $e->getTokenProcessor()->getContextValues('actionSearchResult', 'entityID')
+      + $e->getTokenProcessor()->getContextValues($this->getEntityContextSchema());
+
+    if (!$entityIds) {
+      return NULL;
+    }
+
+    $returnFields = self::getReturnFields($this->activeTokens);
+    if (!in_array('currency', $returnFields)) {
+      // We need currency if we have an amount
+      $returnFields[] = 'currency';
+    }
+    // Get data on all entities for basic and customfield tokens
+    $entities = civicrm_api3('Contribution', 'get', [
+      'id' => ['IN' => $entityIds],
+      'options' => ['limit' => 0],
+      'return' => $returnFields,
+    ]);
+    $prefetch['contribution'] = $entities['values'];
+
+    // Store the activity types if needed
+    if (in_array('payment_instrument', $this->activeTokens)) {
+      $this->paymentInstruments = CRM_Contribute_PseudoConstant::paymentInstrument();
+    }
+
+    // Store the activity statuses if needed
+    if (in_array('status', $this->activeTokens)) {
+      $this->contributionStatuses = CRM_Contribute_BAO_Contribution::buildOptions('contribution_status_id', 'get', []);
+    }
+
+    // Store the campaigns if needed
+    if (in_array('financial_type', $this->activeTokens)) {
+      $this->financialTypes = CRM_Contribute_BAO_Contribution::buildOptions('financial_type_id', 'get', []);;
+    }
+
+    return $prefetch;
+  }
+
+  /**
+   * Evaluate the content of a single token.
+   *
+   * @param \Civi\Token\TokenRow $row
+   *   The record for which we want token values.
+   * @param string $entity
+   *   The name of the token entity.
+   * @param string $field
+   *   The name of the token field.
+   * @param mixed $prefetch
+   *   Any data that was returned by the prefetch().
+   *
+   * @throws \CRM_Core_Exception
+   */
+  public function evaluateToken(TokenRow $row, $entity, $field, $prefetch = NULL) {
+    // Get EntityID either from actionSearchResult (for scheduled reminders) if exists
+    $entityId = $row->context[$this->getEntityContextSchema()];
+
+    $contribution = $prefetch['contribution'][$entityId];
 
     $aliasTokens = $this->getAliasTokens();
     if (in_array($field, ['total_amount', 'fee_amount', 'net_amount'])) {
-      return $row->format('text/plain')->tokens($entity, $field,
-        \CRM_Utils_Money::format($fieldValue, $actionSearchResult->contrib_currency));
+      $row->tokens($entity, $field, \CRM_Utils_Money::format($contribution[$field], $contribution['currency']));
+    }
```

Comment:
Is there metadata that could be used to determine which fields should be formatted as money instead of hard-coding?

Reply:
We probably can read it from the entity "type" field.